### PR TITLE
Support negative `axis` in sparse min/max/argmin/argmax

### LINKED ISSUE #3474: sparse min/max for negative `axis`
* Conditions (you can just paste the output of `python -c 'import cupy; cupy.show_config()'`)
```
CuPy Version          : 8.0.0b3
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 9000
CUDA Driver Version   : 9000
CUDA Runtime Version  : 9000
cuBLAS Version        : 9000
cuFFT Version         : 9000
cuRAND Version        : 9000
cuSOLVER Version      : (9, 0, 0)
cuSPARSE Version      : 9000
NVRTC Version         : (9, 0)
Thrust Version        : 100900
cuDNN Build Version   : 7300
cuDNN Version         : 7300
NCCL Build Version    : 2005
NCCL Runtime Version  : (unknown)
CUB Version           : None
cuTENSOR Version      : None
```

* Code to reproduce

```
cupyx.scipy.sparse.csc_matrix(cupy.arange(9).astype('f').reshape(3, 3) - 4).min(axis=-1)
```

* Error messages, stack traces, or logs

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kataoka/cupy/cupyx/scipy/sparse/data.py", line 284, in min
    return self._min_or_max(axis, out, cupy.min, sum_duplicates, nonzero)
  File "/home/kataoka/cupy/cupyx/scipy/sparse/data.py", line 163, in _min_or_max
    raise ValueError("axis out of range")
ValueError: axis out of range
```
while scipy supports it:
```
>>> scipy.sparse.csc_matrix(np.arange(9).astype('f').reshape(3, 3) - 4).min(axis=-1)
<3x1 sparse matrix of type '<class 'numpy.float32'>'
	with 3 stored elements in COOrdinate format>
```

### PR DESCRIPTION
**Pull request by takagi**
Close #3474.

~~This PR fixes sparse `min` and `max` to take a negative `axis`. I'll parameterize its test in another PR.~~

Merge #3656 first. This PR fixes sparse min/max/argmin/argmax to take a negative `axis`.

### CODE REVIEW

**Comment by takagi** [comment]
I need to fix it to take `axis=None` into account.

**Comment by takagi** [comment]
I've fixed that line to check if `axis` is `None`, but sparse min/max initially does not support `axis=None` as I posted on #3506. Please let me address it as another issue.

File: cupyx/scipy/sparse/data.py [hunk 1]
```diff
@@ -156,11 +156,11 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
 
         validateaxis(axis)
 
-        if axis == 0 or axis == 1:
-            return self._min_or_max_axis(axis, min_or_max, sum_duplicates,
-                                         non_zero)
-        else:
-            raise ValueError("axis out of range")
+        if axis is not None and axis < 0:
+            axis += self.ndim
```
Comment (toslunar) [inline cupyx/scipy/sparse/data.py:160]: spmatrix is always 2-dim.

Comment (takagi) [reply cupyx/scipy/sparse/data.py]: Changed it to a literal!

**Comment by takagi** [comment]
I've rebased this branch to follow #3656, which makes the change much simpler. Please merge #3656 first.

File: cupyx/scipy/sparse/data.py [hunk 2]
```diff
@@ -192,11 +192,11 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
                     assert False
             return m
 
-        if axis == 0 or axis == 1:
-            return self._min_or_max_axis(axis, min_or_max, sum_duplicates,
-                                         non_zero)
-        else:
-            raise ValueError("axis out of range")
```
Comment (toslunar) [inline cupyx/scipy/sparse/data.py:199]: Note: There is `sputils.validateaxis(axis)` above in the method.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py [hunk 3]
```diff
@@ -1058,6 +1057,9 @@ def _make_data_min_nonzero(self, xp, sp, axis):
                 if numpy.isinf(dm_data).all():
                     dm_data[0, 0] = 0
             else:
+                if axis < 0:
+                    axis += 2
+
```
Comment (leofang) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py:1062]: Is this needed?

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csr.py [hunk 4]
```diff
@@ -1248,6 +1246,9 @@ def _make_data_min_nonzero(self, xp, sp, axis):
                 if numpy.isinf(dm_data).all():
                     dm_data[0, 0] = 0
             else:
+                if axis < 0:
+                    axis += 2
+
```
Comment (leofang) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_csr.py:1251]: ditto

**Comment by toslunar** [comment]
Could you resolve the conflicts (with #3676)?

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py (hunk 3 above)
Comment (takagi) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py:1062]: Yes, this is for the branch below:
https://github.com/cupy/cupy/blob/53db8f4047b252176401dc53657286b212fff0e4/tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py#L1067

**Comment by takagi** [comment]
I've resolved the conflict.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**